src/linear_algebra/general_linear_group.lean

```diff
@@ -0,0 +1,269 @@
+units.map_equiv to_lin_alg_equiv'.to_mul_equiv
+
+
+/--Given a matrix with invertible determinant we get an element of `GL n R`-/
+ def GL.mk' (A: matrix n n R) (h: invertible (det A)): GL n R:=
+unit_of_det_invertible A
+
+/--Given a matrix with unit determinant we get an element of `GL n R`-/
+noncomputable def GL.mk'' (A: matrix n n R) (h: is_unit (det A)): GL n R:=
+nonsing_inv_unit A h
+
+
+end
+namespace general_linear_group
+
+
+
+variables {n : Type u} [decidable_eq n] [fintype n] {R : Type v} [comm_ring R]
+
+instance coe_fun : has_coe_to_fun (GL n R) :=
+{ F   := λ _, n → n → R,
+  coe := λ A, A.val }
+
+/--Given a element of `GL n R` it gives the associated linear map-/
+
+def to_lin' (A : GL n R) := matrix.to_lin' A
+
+lemma ext_iff (A B : GL n R) : A = B ↔ (∀ i j, A i j = B i j) :=
+begin
+dsimp at *, fsplit, work_on_goal 0 { intros a i j, induction a, refl }, intros a,
+ ext1, dsimp at *, ext1, solve_by_elim,
+end
+
+@[ext] lemma ext (A B : GL n R) : (∀ i j, A i j = B i j) → A = B :=
+(ext_iff A B).mpr
+
+
+
+section coe_lemmas
+
+variables (A B : GL n R)
+
+
+
+
+
+@[simp] lemma mul_val : ↑(A * B) = A ⬝ B := rfl
+
+@[simp] lemma mul_apply : ⇑(A * B) = (A ⬝ B) := rfl
+
+
+
+
+@[simp] lemma one_val : ↑(1 : GL n R) = (1 : matrix n n R) := rfl
+
+@[simp] lemma one_apply : ⇑(1 : GL n R) = (1 : matrix n n R) := rfl
+
+
+@[simp] lemma to_lin'_mul :
+  to_lin' (A * B) = (to_lin' A).comp (to_lin' B) :=
+matrix.to_lin'_mul A B
+
+@[simp] lemma to_lin'_one :
+  to_lin' (1 : GL n R) = linear_map.id :=
+matrix.to_lin'_one
+
+
+end coe_lemmas
+/-
+
+noncomputable instance :  has_inv (GL n R):=
+⟨λ A, ⟨nonsing_inv A, A, by {simp, apply nonsing_inv_mul, apply GL.is_unit_det _ _ A,},
+by { simp, apply mul_nonsing_inv, apply GL.is_unit_det _ _ A, }, ⟩⟩
+
+@[simp] lemma inv_val : ↑(A⁻¹) = nonsing_inv A := rfl
+
+@[simp] lemma inv_apply : ⇑(A⁻¹) = nonsing_inv A := rfl
+
+
+lemma is_left_inv (A : GL n R): A⁻¹ *  A = 1:=
+
+begin
+ have h0:= GL.is_unit_det _ _ A,
+have:=nonsing_inv_mul A h0, rw ← mul_eq_mul at this, ext, dsimp,rw nonsing_inv,
+cases h0, dsimp at *, solve_by_elim,
+end
+
+
+noncomputable instance : group (GL n R) :=
+{ mul := λ u₁ u₂, ⟨u₁.val * u₂.val, u₂.inv * u₁.inv,
+    by rw [mul_assoc, ← mul_assoc u₂.val, units.val_inv, one_mul, units.val_inv],
+    by rw [mul_assoc, ← mul_assoc u₁.inv, units.inv_val, one_mul, units.inv_val]⟩,
+  one := ⟨1, 1, one_mul 1, one_mul 1⟩,
+  mul_one := λ u,  mul_one u,
+  one_mul := λ u,  one_mul u,
+  mul_assoc := λ u₁ u₂ u₃,  mul_assoc u₁ u₂ u₃,
+  inv := general_linear_group.has_inv.inv,
+  mul_left_inv := λ u,  by {apply is_left_inv,} }
+
+-/
+ instance : group (GL n R) :=infer_instance
+
+/-- Applying `matrix.nonsing_inv` to the coercion is the same as coercing the group inverse. -/
+lemma nonsing_inv_coe_eq_coe_inv (A : GL n R) : (A : matrix n n R)⁻¹ = ↑(A⁻¹) :=
+begin
+  letI := A.invertible,
+  exact (inv_eq_nonsing_inv_of_invertible (A : matrix n n R)).symm,
+end
```

Comment:
There's some weirdness between `↑A` and `⇑A` going on here that I haven't seen before. Asking explicitly for the former fixes it:
```suggestion
lemma nonsing_inv_coe_eq_coe_inv (A : GL n R) : (↑A : matrix n n R)⁻¹ = ↑(A⁻¹) :=
begin
  letI := A.invertible,
  exact (inv_eq_nonsing_inv_of_invertible ↑A).symm,
end
```

Reply:
That works, thank you very much. I've fixed this now and I also re-added the `inv_val` and `inv_apply` lemmas. I'm not sure if they are worth having so I'm happy to remove them if you think its best.